src/test/scala/sigmastate/utils/ByteReaderWriterImpSpecification.scala

```diff
@@ -1,8 +1,53 @@
 package sigmastate.utils
 
-import org.scalatest.{Matchers, PropSpec}
+import org.scalacheck.{Arbitrary, Gen}
 import org.scalatest.prop.PropertyChecks
+import org.scalatest.{Matchers, PropSpec}
+import sigmastate.serialization.Serializer
+import sigmastate.serialization.generators.ValueGenerators
+
+class ByteReaderWriterImpSpecification extends PropSpec
+  with ValueGenerators
+  with PropertyChecks
+  with Matchers {
+
+  private val seqPrimValGen: Gen[Seq[Any]] = for {
+    length <- Gen.chooseNum(1, 100)
+    anyValSeq <- Gen.listOfN(length,
+      Gen.oneOf(
+        Arbitrary.arbByte.arbitrary,
+        Arbitrary.arbShort.arbitrary,
+        Arbitrary.arbInt.arbitrary,
+        Arbitrary.arbLong.arbitrary,
+        arrayGen[Byte]))
+  } yield anyValSeq
+
+  private implicit val arbSeqPrimVal: Arbitrary[Seq[Any]] = Arbitrary(seqPrimValGen)
 
-class ByteReaderWriterImpSpecification extends PropSpec with PropertyChecks with Matchers {
+  property("round trip serialization/deserialization of arbitrary value list") {
+    forAll { values: Seq[Any] =>
+      val writer = values.fold(Serializer.startWriter()) { case (w: ByteArrayWriter, anyVal: Any) =>
+        anyVal match {
+          case v: Byte => w.put(v)
+          case v: Short => w.putShort(v)
+          case v: Int => w.putInt(v)
+          case v: Long => w.putLong(v)
+          case v: Array[Byte] => w.putInt(v.length).putBytes(v)
+          case ref@_ => fail(s"writer: unsupported value type: ${ref.getClass}");
+        }
+      }.asInstanceOf[ByteArrayWriter]
 
+      val reader = Serializer.startReader(writer.toBytes, 0)
+      values.map {
```

Comment:
@greenhat The same here, `map` is not really to organize loops.

Reply:
I've used `map` to collect `Assertion`s for a better error tracing and not to organize loops. However, it does not add a value here since assertion failing exceptions do just fine. Thanks for catching this.